global_body_planner/src/planning_utils.cpp

```diff
@@ -904,79 +956,97 @@ bool isValidState(const State &s, const PlannerConfig &planner_config,
-              << planner_config.terrain.getGroundHeight(s_leg[0], s_leg[1])
-              << std::endl;
-          std::cout << "s.pos\n" << s.pos << std::endl;
-          std::cout << "R_mat.col(0)*x_body\n"
-                    << R_mat.col(0) * x_body << std::endl;
-          std::cout << "R_mat.col(1)*y_body\n"
-                    << R_mat.col(1) * y_body << std::endl;
-          printf("py = %5.3f, %5.3f\n", pitch, yaw);
-          std::cout << "R_mat\n" << R_mat << std::endl;
+      printf("H_MIN exceeded for leg, phase = %d\n", phase);
+      printStateNewline(s);
 #endif
-          return false;
-        }
-      } else if (phase == LEAP_STANCE || phase == LAND_STANCE) {
-        if (leg_height > planner_config.H_MAX) {
+      return false;
+    }
+  }
+
+  // Initialize max_valid_z to infinity (ensures that update will overwrite it)
+  max_valid_z = INFTY;
+
+  // Compute the reachability points in the world frame
+  Eigen::Matrix<double, 3, planner_config.num_reachability_points>
+      reachability_points_world =
+          R_mat * planner_config.reachability_points_body +
+          s.pos.replicate(1, planner_config.num_reachability_points);
+
+  // TODO(jcnorby): allow alternative gaits, check region around foot location
+  for (int i = 0; i < planner_config.num_reachability_points; i++) {
+    Eigen::Vector3d reachability_point = reachability_points_world.col(i);
+    if (!isInMap(reachability_point, planner_config)) {
+#ifdef DEBUG_INVALID_STATE
+      printf("reachability_point not in map, phase = %d\n", phase);
+#endif
+      return false;
+    }
+
+    double reachability_clearance =
+        getZRelToTerrain(reachability_point, planner_config);
+
+    // Check for reachability
+    if (phase == CONNECT) {
+      max_valid_z = std::min(max_valid_z, s.pos[2] + planner_config.H_MAX -
+                                              reachability_clearance);
+      if (reachability_clearance > planner_config.H_MAX) {
+#ifdef DEBUG_INVALID_STATE
+        printf("H_MAX exceeded, phase = %d\n", phase);
+#endif
+        return false;
+      }
+    } else if (phase == LEAP_STANCE) {
+      if (i % 2 == 1) {
+        max_valid_z = std::min(max_valid_z, s.pos[2] + planner_config.H_MAX -
+                                                reachability_clearance);
+        if (reachability_clearance > planner_config.H_MAX) {
 #ifdef DEBUG_INVALID_STATE
           printf("H_MAX exceeded, phase = %d\n", phase);
           printStateNewline(s);
 #endif
           return false;
         }
       }
-    }
-  }
-
-  // Check the center of the underside of the robot for penetration
-  double height =
-      (s.pos[2] + R_mat(2, 2) * z_body) -
-      planner_config.terrain.getGroundHeight(s.pos[0] + R_mat(0, 2) * z_body,
-                                             s.pos[1] + R_mat(1, 2) * z_body);
-  if (height < planner_config.H_MIN) {
+    } else if (phase == LAND_STANCE) {
+      // if (i % 2 == 0) {
```

Comment:
remove?

Reply:
Leaving this in since I'm still tweaking but its clearer now.